src/devices/VirtualGpio/README.md

```diff
@@ -0,0 +1,45 @@
+# Virtual GPIO Controller
+
+As the name says, `VirtualGpio` is a virtualized GPIO controller. It based on `GpioController` and defines an input method.
+
+It allows you to control input pins of GPIO using software code without having to rely on specific hardware. For simulation, testing, etc.
+
+# Usage
+
+## Create vGPIO controller
+```csharp
+VirtualGpioController vGpio = VirtualGpioController.Create(16);
+```
+
+## Add event handlers
+```csharp
+PinChangeEventHandler pinValueChanged = (sender, args) =>
+{
+    Console.WriteLine($"PinNumber: {args.PinNumber}, ChangeType: {args.ChangeType}");
+};
+
+vGpio.OutputPinValueChanged += pinValueChanged;
+vGpio.InputPinValueChanged += pinValueChanged;
+```
+
+## Write output pins
+```csharp
+vGpio.OpenPin(0, PinMode.Output);
+vGpio.Write(0, PinValue.High); // set to high -> trigger `OutputPinValueChanged` event
+```
+
+## Control input pins
+```csharp
+vGpio.Input(0, PinValue.Low); // connect to low -> shorted, throw exception
+```
```

Comment:
You need to add text between the code blocks. Also please respect the new lines you need to have.